### PR TITLE
feat(crons): Use an in memory lock to process check-ins

### PR DESCRIPTION
**Pull request by evanpurkhiser**
As part of https://github.com/getsentry/sentry/issues/60292 We need to
ensure that check-ins for a monitor are processed in order when
multi-threading. We can do this by holding a lock on the check-in
processing execution for each monitor slug.

This does NOT compeltely gaurentee in-order processing when using
multi-threading, it is still possible that two check-ins may be queued
into threads and start out of order, where the later check-in takes the
lock.

This is especially true since there is a small amount of work that needs
to be done before we can hold the lock (JSON parsing of payload).

### CODE REVIEW

File: src/sentry/monitors/consumers/monitor_consumer.py [hunk 1]
```diff
@@ -59,6 +61,15 @@
 LOCK_EXP_BASE = 2.0
 
 
+# Use the lru_cache to keep our per-monitor-locks around in a thread safe
+# manor. We'll use this to ensure that we process checkins for a individual
+# monitor_environment in sequeunce, even though we're multi-threading our
+# consumer.
+@lru_cache(maxsize=100_000, ttl=60)
+def get_monitor_lock(kry: str):
+    return Lock()
```
Comment (evanpurkhiser) [inline src/sentry/monitors/consumers/monitor_consumer.py:70]: Not sure how safe this actually is. But seems like a decent way to hold a mapping of `key → lock` with a TTL expiration.

Though I think once we un-lock we would ideally remove the lock, so maybe this isn't the perfect solution

File: src/sentry/monitors/consumers/monitor_consumer.py [hunk 2]
```diff
@@ -603,7 +614,19 @@ def _process_message(
...
+        slug = params.get("monitor_slug")
+        environment = params.get("environment", "none")
+
+        # We lock the check-in processing for the monitor environment to ensure
+        # our multi-threading consumer processes check-ins in order.
+        #
+        # XXX(epurkhiser): This may be a best-effort lock. It is possible one
+        # thread processes a later check-in before another. Luckily most
+        # open/close check-ins will have a decent gap in time between them.
+        lock = get_monitor_lock(f"{slug}-{environment}")
```
Comment (rjo100) [inline src/sentry/monitors/consumers/monitor_consumer.py:626]: don't we want to scope this per project/org as well?

File: src/sentry/monitors/consumers/monitor_consumer.py [hunk 3]
```diff
@@ -59,6 +61,15 @@
 LOCK_EXP_BASE = 2.0
 
 
+# Use the ttl_cache to keep our per-monitor-locks around in a thread safe
+# manor. We'll use this to ensure that we process checkins for a individual
+# monitor_environment in sequeunce, even though we're multi-threading our
```
Comment (rjo100) [inline src/sentry/monitors/consumers/monitor_consumer.py:66]:
```suggestion
# manner. We'll use this to ensure that we process check-ins for a individual
# monitor_environment in sequence, even though we're multi-threading our
```

File: src/sentry/monitors/consumers/monitor_consumer.py (hunk 1 above)
Comment (rjo100) [reply src/sentry/monitors/consumers/monitor_consumer.py]: yeah don't we need to `acquire` the lock as well?

**Comment by evanpurkhiser** [comment]
Going to take a different approach here to performance. See https://github.com/getsentry/sentry/issues/60292

**Pull request closed without merging**